Creating a buffer for parseHeader and parsePostData

patrickjahns opened the pull request:
This addresses #615 and also prepares for working on multipart requests

The previous implementation did not account that header information could be split between several packets. Also post data can be split across several TCP packets.

The changes address this by storing not processed data in a temporary buffer and pre pending the new tcp buffer so it will be parsed together

Has been tested with several requests (GET/POST including Cookies and large headers)

patrickjahns commented:
@hreintke 

any comments so far?

PTDreamer commented on Sming/SmingCore/Network/HttpRequest.cpp:
```diff
@@ -82,77 +83,94 @@ String HttpRequest::getContentType()
 	{
-		nextLine = NetUtils::pbufFindStr(buf, "\r\n", line);
+
+		nextLine = tmpbuf.indexOf("\r\n", line+2);
```
Shouldn't the last parenthesis be placed before the "+" operator?

patrickjahns replied:
I`ll verify it again

PTDreamer commented on Sming/SmingCore/Network/HttpRequest.cpp:
```diff
@@ -82,77 +83,94 @@ String HttpRequest::getContentType()
 		return eHPR_Successful;
 	}
+	headerDataProcessed += buf->tot_len;
+	tmpbuf = tmpbuf.substring(line, buf->tot_len);
```
Shouldn't the size be "buf->tot_len - line" instead of "but->tot_len" ?

patrickjahns replied:
In regards to  the size see the talk on substring below. But you caught something else - it should
`tmpbuf.substring(line, tmpbuf.length()` - the length of the tmpbuf can be longer than buf->tot_len

totally missed that

PTDreamer commented on Sming/SmingCore/Network/HttpRequest.h:
```diff
@@ -47,20 +47,21 @@ class HttpRequest
 	HashMap<String, String> *cookies;
 	int postDataProcessed;
-	bool combinePostFrag;
+	int headerDataProcessed;
```
headerDataProcessedLenght would be closer to the truth

PTDreamer commented on Sming/SmingCore/Network/HttpRequest.cpp:
```diff
@@ -165,61 +183,35 @@ HttpParseResult HttpRequest::parsePostData(HttpServer *server, pbuf* buf)
 
-		String ItemName = NetUtils::pbufStrCopy(buf, startItem, delimItem - startItem);
-		String ItemValue = NetUtils::pbufStrCopy(buf, delimItem + 1, nextItem - delimItem - 1);
+		String ItemName = buf.substring(startItem, delimItem);
```
Shouldn't it be buf.substring(startItem, delimItem - startitem) ?

PTDreamer commented on Sming/SmingCore/Network/HttpRequest.cpp:
```diff
@@ -165,61 +183,35 @@ HttpParseResult HttpRequest::parsePostData(HttpServer *server, pbuf* buf)
-		String ItemName = NetUtils::pbufStrCopy(buf, startItem, delimItem - startItem);
-		String ItemValue = NetUtils::pbufStrCopy(buf, delimItem + 1, nextItem - delimItem - 1);
+		String ItemName = buf.substring(startItem, delimItem);
+		String ItemValue = buf.substring(delimItem+1, nextItem);
```
Shouldn't it be buf.substring(delimItem + 1, nextItem - delimItem - 1) ?

patrickjahns replied:
Short answer since I am on mobile. Substring uses absolute numbers
substring(Startposition, endposition)  while pbuf used (Startposition,
length). So this should be correct
On Mar 1, 2016 16:10, "Jose Barros" <notifications@github.com> wrote:

> In Sming/SmingCore/Network/HttpRequest.cpp
> <https://github.com/SmingHub/Sming/pull/622#discussion_r54579516>:
>
> >
> > -		String ItemName = NetUtils::pbufStrCopy(buf, startItem, delimItem - startItem);
> > -		String ItemValue = NetUtils::pbufStrCopy(buf, delimItem + 1, nextItem - delimItem - 1);
> > +		String ItemName = buf.substring(startItem, delimItem);
> > +		String ItemValue = buf.substring(delimItem+1, nextItem);
>
> Shouldn't it be buf.substring(delimItem + 1, nextItem - delimItem - 1) ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SmingHub/Sming/pull/622/files#r54579516>.
>

PTDreamer replied:
Wow this is confusing so this string class is not the same as defined in the c++ std namespace? std::string::substr arguments are to size_t's with the first being start index and the second being length.

PTDreamer replied:
Found the definition here https://github.com/SmingHub/Sming/blob/develop/Sming/Wiring/WString.cpp#L586 and of course you are right. However even if not related to this PR I find this kind of subtleties counter intuitive and should be avoided by not using the same name for functions which exist in "standard c++".

patrickjahns replied:
@PTDreamer 
This is related to SMING being compatible to Arduino framework - see their Substring implementation here: https://www.arduino.cc/en/Reference/StringSubstring

PTDreamer commented:
No comments regarding coding style since I found no reference to the style adopted by this project. Linux Kernel style is the most common for this kind of projects and this PR doesn't conform in several places.
I don't fully understand the idea behind processing data chunks as they come instead of just increment (append data to) a buffer  waiting for a complete header and process that. It would simplify code and increase readability (IMHO) without any performance impact.
PS: Don't know if my comments are welcomed or not since I haven't contributed with any code to this repo/project, but honestly I had nothing better to do :).

patrickjahns commented:
@PTDreamer 

See the Summary on the first page - there is the compatibility listed to arduino libraries ...

In regards to processing packets as they arrive (actually you can relate to this as a concept of processing a stream) - it is more memory friendly. While header data might be "small" (compared to post requests) - it requires more ram than just keeping unprocessed packet data in a buffer.

Before this PR there was no processing of headers split accross packets - (see #615 )
This is also meant as basis for processing multipart/post requests - a feature which I am working on in a different PR. While your idea of processing the header at once works for the header - it will not properly work for POST bodies - especially not for multipart requests - latest then I would need to implement the concept - and why not process both the same way? It is more intuitive than to process header and post requests differently

PTDreamer commented:
> See the Summary on the first page - there is the compatibility listed to arduino libraries ...
> Any further discussion in regard to this should rather be in a different ISSUE - to keep things related to this PR

Not sure if the comment is related to the coding style or discussion about the substring function.


> it is more memory friendly. While header data might be "small" (compared to post requests) - it requires more ram than just keeping unprocessed packet data in a buffer

It is not obvious to me how that is true. You are trading raw data buffer space with hash stored values which looking at the implementation bloats the amount of memory needed to store the data (which is true for all hash implementations).

> Before this PR there was no processing of headers split accross packets - (see #615 )

No argument there. It is obvious there was a bug, the data will need to be appended to a buffer between callback calls or it will get ditched.

> While your idea of processing the header at once works for the header - it will not properly work for POST bodies

Why not? You can just buffer the stream until you find "delimiter "--"" and parse it then.

In the end your code should work fine (although I still think you have a misplaced parenthesis), it is just a matter of personal coding style. Long do/whiles and lots of global variables are not my thing, also I found this PR hard to read not only by the code style but also by the variable naming.
Good job though, I hope I can find something to contribute with so you can take a turn at bashing my code.

patrickjahns commented:
> Not sure if the comment is related to the coding style or discussion about the substring function.

It was related to the discussion in regards to substring and compatibility with arduino libraries. And as suggested - i think this is best suited to be discussed in it`s own issue since it is not related to this PR

> It is not obvious to me how that is true. You are trading raw data buffer space with hash stored values which looking at the implementation bloats the amount of memory needed to store the data (which is true for all hash implementations).

You need to differentiate between the buffer and the Hashmap - the Hashmap implementation was already in use even before I suggested the buffer. When keeping everything in one buffer before parsing, you would need the space for the buffer and the hashmap at the same time. Parsing as far as it goes, the buffer data needed can be smaller and leave room for the final hashmap to be worked with. 

> Why not? You can just buffer the stream until you find "delimiter "--"" and parse it then.

Since multipart requests are not only variables/fields to be kept in memory, but also can be files - I don`t see your suggestion viable - it would need again 2 different implementations for one task. For suggestions/discussion on multipart and file uploads please join #552 

> In the end your code should work fine (although I still think you have a misplaced parenthesis), it is just a matter of personal coding style. Long do/whiles and lots of global variables are not my thing, also I found this PR hard to read not only by the code style but also by the variable naming.

Please compare the version before and after my PR - I based my solution on the already exiting coding style. In what way would you improve the current implementation ? 

> Good job though, I hope I can find something to contribute with so you can take a turn at bashing my code.

I hope you don`t see this discussion as a need for bashing code - as long as it improves code and usability in the end - it`s good to discuss it. And I am not too well suited for doing c/c++ code - I am more home with python. But since I am working on a project with an esp - I might as well help to improve it right? ;-)

hreintke commented:
@patrickjahns : 
>In the end your code should work fine (although I still think you have a misplaced parenthesis)
Is there still an update needed from your code.?

On In the PR now there is 
> 	Merge pull request #1 from SmingHub/develop … 
That contains (part of ?) another PR. 
I am not the most experienced in git so am not sure how that effects the merging.
Any change that you can update the PR for that ?

patrickjahns commented:
@hreintke 

I had trouble squashing - will try again later and update accordingly. Please wait with merging until then. I still want to verify the part with the misplaced parenthesis - didn`t get around to it until now

hreintke commented:
OK Clear

patrickjahns commented:
@PTDreamer 
Thanks for the hint with the line numbers - the +2 calculation was unnecessary and was a leftover from previous testing. Removed and now the PR is clean

@hreintke 
squashed and ready to merge - I can prepare another PR for RTOS then

patrickjahns commented:
@avr39-ripe 
Mentioned that query parameters are not parsed/parsed correctly - will need to have another look

avr39-ripe commented:
Nither queryParameter nor urlencoded form data DO NOT work ;(( switching to regular sming/develop solve the problem.. and introduce new one with packet hassle

patrickjahns commented:
@avr39-ripe 

pushed the proposed fix - please test and confirm that query and post parameters are working now as expected (please test only this branch and don`t include rawbody yet)

avr39-ripe commented:
cherry-pick last two commits on top of develop and it works. tested by using response.getQueryParameter()

patrickjahns commented:
I`d say this is ready to merge - maybe someone else can also confirm ?

avr39-ripe commented:
If @dobrishinov try this and confirm.. He expect problems with split/truncated data.. but it needs rawBody fix also..
